results/cmd/api/main.go

```diff
@@ -119,19 +120,49 @@ func (s *server) GetTaskRunResult(ctx context.Context, req *pb.GetTaskRunRequest
 
 // UpdateTaskRun receives TaskRun and FieldMask from client and uses them to update records in local Sqlite Server.
 func (s *server) UpdateTaskRunResult(ctx context.Context, req *pb.UpdateTaskRunRequest) (*pb.TaskRunResult, error) {
-	// Update the entire row in database based on uid of taskrun.
-	statement, err := s.db.Prepare("UPDATE taskrun SET name = ?, namespace = ?, taskrunlog = ? WHERE results_id = ?")
+	// Find corresponding TaskRun in the database according to results_id.
+	resultsID, err := uuid.Parse(req.GetResultsId())
 	if err != nil {
-		log.Printf("failed to update a existing taskrun: %v\n", err)
-		return nil, fmt.Errorf("failed to update a exsiting taskrun: %w", err)
+		log.Fatal("failed to parse resultID string into resultsID UUID", err)
+		return nil, fmt.Errorf("failed to find a taskrun: %w", err)
+	}
+	rows, err := s.db.Query("SELECT taskrunlog FROM taskrun WHERE results_id = ?", resultsID)
+	if err != nil {
+		log.Fatalf("failed to query on database: %v", err)
+		return nil, fmt.Errorf("failed to query on a taskrun: %w", err)
 	}
+	taskrun := &pb.TaskRun{}
+	rowNum := 0
+	for rows.Next() {
+		var taskrunblob []byte
+		rowNum++
+		if rowNum >= 2 {
+			log.Println("Warning: multiple rows found")
+			break
+		}
+		rows.Scan(&taskrunblob)
+		if err := proto.Unmarshal(taskrunblob, taskrun); err != nil {
+			log.Fatal("unmarshaling error: ", err)
+			return nil, fmt.Errorf("failed to unmarshal taskrun: %w", err)
+		}
+	}
+	// Merge TaskRun from client into existing TaskRun based on fieldmask.
 	taskrunFromClient := req.GetTaskRun()
-	blobData, err := proto.Marshal(taskrunFromClient)
+	fieldMask := req.GetFieldMask()
+	msk, _ := mask.FromFieldMask(fieldMask, taskrun, false, false)
+	msk.Merge(taskrunFromClient, taskrun)
```

Comment:
It's hard to know from the variable names what's the purpose of each. I think `msk.Merge(taskrunFromClient, update)` might be clearer.

Reply:
done